### PR TITLE
1436517: Fix api base detection for rhevm version 3 and 4

### PR DESCRIPTION
**Pull request by cnsnyder**
Please ask me for configuration to test this PR.

This PR allows the rhevm backend to check two possible api base urls then use the one that was successful for subsequent calls to the api. In addition, this PR allows the rhevm backend to continue to report on hypervisors even if the sockets can not be collected for a given vm.

### CODE REVIEW

**Comment by cnsnyder** [comment]
/me goes to fix broken tests, thanks for the catch travis

File: virtwho/virt/rhevm/rhevm.py [hunk 1]
```diff
@@ -109,16 +105,20 @@ def get_version(self):
         try:
             headers = dict()
             headers['Version'] = '3'
-            response = requests.get(urlparse.urljoin(self.url, 'api'),
+            # We will store the api_base that seems to work and use that for future requests
+            base = 'api'
```
Comment (kahowell) [inline virtwho/virt/rhevm/rhevm.py]: This line can go away maybe...

File: virtwho/virt/rhevm/rhevm.py [hunk 2]
```diff
@@ -60,6 +60,7 @@ def __init__(self, logger, config, dest, terminate_event=None,
                                     interval=interval,
                                     oneshot=oneshot)
         self.url = self.config.server
+        self.api_base = 'api/'
```
Comment (kahowell) [inline virtwho/virt/rhevm/rhevm.py]: I think you should leave trailing slash off of `self.api_base`, and instead put `/` as a prefix on `{clusters,hosts,vms}_endpoint`...

File: virtwho/virt/rhevm/rhevm.py [hunk 3]
```diff
@@ -109,16 +105,20 @@ def get_version(self):
...
+            # We will store the api_base that seems to work and use that for future requests
+            base = 'api'
+            response = requests.get(urlparse.urljoin(self.url, base),
                                     auth=self.auth,
                                     headers=headers,
                                     verify=False)
-            if response.status_code == 404 and 'ovirt-engine' not in self.url:
-                response = requests.get(urlparse.urljoin(self.url, 'ovirt-engine/api'),
+            if response.status_code == 404:
+                base = 'ovirt-engine/api'
```
Comment (kahowell) [inline virtwho/virt/rhevm/rhevm.py]: I would just go ahead and assign `self.api_base` here instead.

File: virtwho/virt/rhevm/rhevm.py [hunk 4]
```diff
@@ -109,16 +105,20 @@ def get_version(self):
...
+                response = requests.get(urlparse.urljoin(self.url, base),
                                         auth=self.auth,
                                         headers=headers,
                                         verify=False)
             response.raise_for_status()
+            self.api_base = base +'/'
```
Comment (kahowell) [inline virtwho/virt/rhevm/rhevm.py]: This line goes away if you follow my suggestions.

File: virtwho/virt/rhevm/rhevm.py [hunk 5]
```diff
@@ -109,16 +105,20 @@ def get_version(self):
         try:
             headers = dict()
             headers['Version'] = '3'
-            response = requests.get(urlparse.urljoin(self.url, 'api'),
+            # We will store the api_base that seems to work and use that for future requests
+            base = 'api'
+            response = requests.get(urlparse.urljoin(self.url, base),
```
Comment (kahowell) [inline virtwho/virt/rhevm/rhevm.py]: I would just go ahead and use `self.api_base` here.

**Comment by cnsnyder** [comment]
@kahowell PR updated according to requests. Please check again to be sure I did not miss anything. Thank you for the review!

**Comment by cnsnyder** [comment]
[1436517: Fix 404 on RHEV detection](https://trello.com/c/TuE79Tfg/362-1436517-fix-404-on-rhev-detection)

**Pull request merged**